### ISSUE TITLE
Questions about the codes

### ISSUE THREAD

**Issue by SMZCC** [liulj13/pyMDNet-VOT2OTB#1]
Hello, I'm glad to read your codes, thx.
However, I can't understand this part in codes:
```python
class LRN(nn.Module):
    def __init__(self):
        super(LRN, self).__init__()

    def forward(self, x):
        #
        # x: N x C x H x W
        pad = Variable(x.data.new(x.size(0), 1, 1, x.size(2), x.size(3)).zero_())
        x_sq = (x**2).unsqueeze(dim=1)
        x_tile = torch.cat((torch.cat((x_sq,pad,pad,pad,pad),2),
                            torch.cat((pad,x_sq,pad,pad,pad),2),
                            torch.cat((pad,pad,x_sq,pad,pad),2),
                            torch.cat((pad,pad,pad,x_sq,pad),2),
                            torch.cat((pad,pad,pad,pad,x_sq),2)),1)
        x_sumsq = x_tile.sum(dim=1).squeeze(dim=1)[:,2:-2,:,:]
        x = x / ((2.+0.0001*x_sumsq)**0.75)
        return x
```
I know it's the implemention of LRN ,but I can't understand how it does!
Especially, here:
```python
 x_sq = (x**2).unsqueeze(dim=1)
        x_tile = torch.cat((torch.cat((x_sq,pad,pad,pad,pad),2),
                            torch.cat((pad,x_sq,pad,pad,pad),2),
                            torch.cat((pad,pad,x_sq,pad,pad),2),
                            torch.cat((pad,pad,pad,x_sq,pad),2),
                            torch.cat((pad,pad,pad,pad,x_sq),2)),1)
        x_sumsq = x_tile.sum(dim=1).squeeze(dim=1)[:,2:-2,:,:]
```
- why should it be unsqueezed?
- how do several torch.cat work? that's so abstract  almostly making me mad
- at last, how do I know which dim I should sum and squeeze ? 
Would you please teach me? 
Being looking forward to your reply with best regards

**Comment by liulj13** [comment]
Actually, you can debug it and see the intermediate results to help you understand the process.
For example, the x is 1x96x51x51, then the pad is defined as 1x1x1x51x51, x_sq is 1x1x96x51x51, 
so the result of (torch.cat((x_sq,pad,pad,pad,pad),2) is 1x1x(96+1+1+1+1)x51x51, so x_tile is 1x(1+1+1+1+1)x100x51x51. Therefore, the x_tile.sum(dim=1) is 1x1x100x51x51, and x_tile.sum(dim=1).squeeze(dim=1) is 1x100x51x51, and x_sumsq is 1x96x51x51 which is the same with x.

**Comment by SMZCC** [comment]
#### Thank you for your reply
- I have no problems with the calculation of  the dimentions 
- what confuses me is that : why doing like this can implement the LRN ? In another word ,  how can I rectify the correctness of the result? Or ,when in programing ,we only need to think about the consistence of the dimentions of the input and the output without thinking about  whether the number of the matrix put in the right place(because of  too much dimentions ,I can't write them on the paper to check ), so this is where the question comes.
#### Best regards

**Comment by liulj13** [comment]
LRN is to normalize the value in a local area, and in this case, the local area is the 4 neighbours of the point in the 96-dimentional feature. More precisely, the sum process sums the x_{i-2}, x_{i-1}, x_i, x_{i+1} and x_{i+2} for each x_i in the 96-dimentional feature, because the x_sq is shifted and sumed after the torch.cat process. Hope it can solve your problem.

**Comment by SMZCC** [comment]
#### Thank you for your patience
- I know how the LRN works(only in text,not in programmig), maybe, I didn't express clearly, sorry, however, what still suffers me is that: 
   + why the second parameter of`torch.cat((x_sq,pad,pad,pad,pad),2)` is 2, not 1 or anything others. (maybe your answer is that that dimention means channel) while  the following is 1 (the last parameter)
         ```python
             torch.cat((torch.cat((x_sq,pad,pad,pad,pad),2),  
                            torch.cat((pad,x_sq,pad,pad,pad),2),  
                            torch.cat((pad,pad,x_sq,pad,pad),2),  
                            torch.cat((pad,pad,pad,x_sq,pad),2),  
                            torch.cat((pad,pad,pad,pad,x_sq),2)),1)
         ```


   + why should the x_sq shift in turn ?
        + torch.cat((x_sq, pad, pad, pad, pad), 2), x_sq  is the shape of (num, 1, channels, height, width)
               the pad is the shape of (1, 1, 1, height, width), I can't imagine how does this operate help calculate the sum of the 5 channels.
        + if I implemented the LRN, I would use a `for` to egordic each channel then use indices to get each number of the five channels and add them which is intuitive and simple ,so , I can't imagine why the `torch.cat(torch.cat(),torch.cat(),torch.cat(),torch.cat(), torch.cat())` does the same thing.
    + how do you  know you should sum which dim?

##### Now, I have a dim feeling of the problem, maybe the problem itself is not directed against the implementation of  the LRN ,but more about programming thinking.

#### Best regards

**Comment by liulj13** [comment]
I will give you a more concrete example to help you understand that, x_sq is 1x1x96x51x51, and we need to normalize it. First, you can neglect the 51x51, because the normalize is for the channel dimention, so we extract one from the 51x51 plane and we have a vector 1x96. Then we normalize it by padding and shifting it and summing it. So we have
x_1  x_2  x_3  x_4  **x_5**  x_6  ....................................   x_93   x_94  x_95  x_96 0       0       0        0 
0     x_1  x_2  x_3  **x_4**  x_5  x_6 .............................   x_92  x_93   x_94  x_95 x_96  0       0        0 
0     0     x_1  x_2  **x_3**  x_4  x_5  x_6 .....................  x_91  x_92   x_93  x_94 x_95   x_96  0        0
0     0     0     x_1  **x_2**  x_3  x_4  x_5  x_6 .............  x_90  x_91  x_92   x_93  x_94  x_95  x_96   0
0     0     0     0     **x_1**  x_2  x_3  x_4  x_5  x_6 ....  x_89   x_90  x_91  x_92  x_93  x_94  x_95   x_96 
and you sum them as column and you can get the normalize result.
The inside cat of torch.cat(torch.cat(),torch.cat(),torch.cat(),torch.cat(), torch.cat()) forms each of the 5 rows, and the outside cat combines the 5 rows.

**Comment by SMZCC** [comment]
#### Thank you 
##### It's much clear for what has happened.
##### Good job 👍 
- However, when the shape of the x_q is` (n, height, width, channels)`, do we have to change the dimentions from `(n, height, width, channels)` to `(n, channels, height, width)` ?, if needn't , how to make it?
- In this problem,  we only add planes, while the matrix to add with a much higher dimentions(for example, it's  a x b x c instead of height x width) , can we directly sum the columns(maybe can't ,then how to deal with it)?
#### Best regards

**Comment by liulj13** [comment]
For the first problem, the location of the dimentions is equal, and you only have to pad and sum the corresponding dimention of channels, so you do not need to change it.
For the second problem, becasue the sum of column is for the channel, no matter it is 2 dims or 3 dims, we only pad and sum the 96-dim feature. The only difference is the plane becomes cube but the pad and sum is for the channel, so it is totally the same code.

**Comment by SMZCC** [comment]
#### Thank you very much.
#### Best regards
😄
